### PR TITLE
cdrom: Return index 1/start of track data instead of index 0/start of pregap data

### PR DESCRIPTION
**Pull request by 987123879113**
Stumbled on a related bug with the way pregap is added to the TOC while trying to fix CDDA audio issues in Firebeat and System 573.

MAME vs Duckstation (both using the same CHD):
image

Results on real PS1:
image

Results with this patch applied:
image

I came up with a simple test program that uses `CdGetToc` and `CdPosToInt` to display the reported MSF and LBA values. Although this change will affect everything that uses cdrom.cpp, I chose the PS1 emulator for testing because I am able to reproduce the bug I'm hunting down using the PS1's CD player functionality in the BIOS.

Testing method for MAME and Duckstation:
1) Load test program cdtester.chd
2) Change disc using file menu from cdtester.chd to a different CHD, in this case a custom CHD named 845jba02.chd
3) Press button which calls `CdGetToc` and `CdPosToInt` on the currently inserted CD and display the results

Same for real PS1 except the CD is switched out for step 2.

845jba02.chd is a CHD made from a CloneCD dump converted to bin/toc. This is what the actual track listing looks like according to the CHD:
```
chdman - MAME Compressed Hunks of Data (CHD) manager 0.229 (unknown)
Input file:   /home/otobear/mame/roms/ddrj/845jba02.chd
File Version: 5
Logical size: 352,727,424 bytes
Hunk Size:    19,584 bytes
Total Hunks:  18,011
Unit Size:    2,448 bytes
Total Units:  144,088
Compression:  cdlz (CD LZMA), cdzl (CD Deflate), cdfl (CD FLAC)
CHD size:     217,136,744 bytes
Ratio:        61.6%
SHA1:         b4dec83e74060f794bff090f6f464cb1d5f51191
Data SHA1:    62d897bb202f16d7ec413845310732b81e1c04b4
Metadata:     Tag='CHT2'  Index=0  Length=93 bytes
              TRACK:1 TYPE:MODE1_RAW SUBTYPE:RW_RAW FRAMES:8661 PREGAP:0 P
Metadata:     Tag='CHT2'  Index=1  Length=91 bytes
              TRACK:2 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:2700 PREGAP:150 PGT
Metadata:     Tag='CHT2'  Index=2  Length=89 bytes
              TRACK:3 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:1287 PREGAP:0 PGTYP
Metadata:     Tag='CHT2'  Index=3  Length=89 bytes
              TRACK:4 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:4621 PREGAP:0 PGTYP
Metadata:     Tag='CHT2'  Index=4  Length=89 bytes
              TRACK:5 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:5661 PREGAP:0 PGTYP
Metadata:     Tag='CHT2'  Index=5  Length=89 bytes
              TRACK:6 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:3762 PREGAP:0 PGTYP
Metadata:     Tag='CHT2'  Index=6  Length=89 bytes
              TRACK:7 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:1801 PREGAP:0 PGTYP
Metadata:     Tag='CHT2'  Index=7  Length=89 bytes
              TRACK:8 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:1802 PREGAP:0 PGTYP
Metadata:     Tag='CHT2'  Index=8  Length=89 bytes
              TRACK:9 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:1801 PREGAP:0 PGTYP
Metadata:     Tag='CHT2'  Index=9  Length=90 bytes
              TRACK:10 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:1803 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=10  Length=90 bytes
              TRACK:11 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:1805 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=11  Length=90 bytes
              TRACK:12 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:1801 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=12  Length=90 bytes
              TRACK:13 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:1801 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=13  Length=90 bytes
              TRACK:14 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:1803 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=14  Length=90 bytes
              TRACK:15 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:1800 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=15  Length=90 bytes
              TRACK:16 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:1801 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=16  Length=90 bytes
              TRACK:17 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:1801 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=17  Length=90 bytes
              TRACK:18 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:7081 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=18  Length=90 bytes
              TRACK:19 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:7351 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=19  Length=90 bytes
              TRACK:20 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:7836 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=20  Length=90 bytes
              TRACK:21 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:7867 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=21  Length=90 bytes
              TRACK:22 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:7075 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=22  Length=90 bytes
              TRACK:23 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:7034 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=23  Length=90 bytes
              TRACK:24 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:7873 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=24  Length=90 bytes
              TRACK:25 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:7860 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=25  Length=90 bytes
              TRACK:26 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:6901 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=26  Length=90 bytes
              TRACK:27 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:6806 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=27  Length=90 bytes
              TRACK:28 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:7754 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=28  Length=90 bytes
              TRACK:29 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:6374 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=29  Length=90 bytes
              TRACK:30 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:3709 PREGAP:0 PGTY
Metadata:     Tag='CHT2'  Index=30  Length=90 bytes
              TRACK:31 TYPE:AUDIO SUBTYPE:RW_RAW FRAMES:5995 PREGAP:0 PGTY
```

### CODE REVIEW

**Comment by 987123879113** [comment]
Advice is requested if anyone more knowledgeable about this part of the code sees this. The exact bug I'm hunting down is still not solved by this PR but I believe it is related. 

The exact issue is: track 2 (the first audio track after the data track in track 1) is ~2 seconds delayed seemingly due to the pregap shown in the chdman info listing. If you load the CD player in the PS1 BIOS and play track 2, it will play roughly at 2 seconds into the track. Using Duckstation, the same exact CHD works as expected with track 2 playing immediately. So I think there is more to this bug somewhere either in the cdrom code or possibly chdcd but I am not sure exactly where.

**Comment by rb6502** [comment]
Check the actual CD-ROM device being used and see how it figures out the LBA to start at on a play audio command.

**Comment by 987123879113** [comment]
@rb6502 The PS1 emulator and the Firebeat/System 573 code have two completely different ways of handling CD drive stuff. 

For Firebeat and System 573 it ends up going through [src/devices/machine/t10mmc.cpp](https://github.com/mamedev/mame/blob/master/src/devices/machine/t10mmc.cpp). It does a `T10MMC_CMD_READ_TOC_PMA_ATIP` command and then uses the returned LBAs to request specific tracks to be played through `T10MMC_CMD_PLAY_AUDIO_12`. 

The PS1 emulator uses `psxcd_device` from [src/mame/machine/psxcd.cpp](https://github.com/mamedev/mame/blob/master/src/mame/machine/psxcd.cpp) which is its own machine and implements its own CD drive entirely separate from the ATAPI CD-ROM drive/t10mmc stuff used in Firebeat and System 573. Furthermore, CDDA audio playback stuff seems to be implemented in the PSX SPU code ([src/devices/sound/spu.cpp](https://github.com/mamedev/mame/blob/master/[src/devices/sound/spu.cpp)), so it's also separate from [src/devices/sound/cdda.cpp](https://github.com/mamedev/mame/blob/master/src/devices/sound/cdda.cpp) which is used by t10mmc.cpp.

The issue occurs between multiple drivers using multiple different CD-ROM and CDDA implementations so I'm thinking that rules out a lot of things. And since Duckstation is able to load and play the CHD properly, I am thinking it's possibly a bug when loading the CHD somewhere.

**Comment by 987123879113** [comment]
I've attached a CHD (only contains a dummy file + sine and sawtooth audio tracks) that can reproduce both the pregap issue and the ~2 second delay issue mentioned in this PR.

image

Example video: https://streamable.com/lkfn0m

[mydisc2.zip](https://github.com/mamedev/mame/files/6177588/mydisc2.zip)

**Comment by rb6502** [comment]
Right, I wasn't saying the problem was necessarily in that specific device, just that that's where I'd start tracing this problem.  It's getting the wrong values from somewhere.

**Comment by DavidHaywood** [comment]
as mentioned before, .cue is one of the worst 'non-standards' around, different programs ended up with different interpretations and layers of hacks, others followed suite.  Quite often the interpretation of the cue, based on the raw data on the disc ends up entirely down to the drive in question too, with many opposing each other, or 'correcting' what they see when they shouldn't (3 second pregaps on some consoles that get seen as 2 second on PC drives for example, because it assumes it's an error)  You're probably going to find a lot of MAME drivers with kludges for this reason, because the collections converted adhered to one standard - sometimes how a specific PC drive saw things and how a specific piece of software generated that cue.

Unless you look at the duckstation code, I don't really know what it's doing.

Unfortunately what we need is a *real* CD dumping technique, as exists for floppies, but so far every person I've heard of looking at CDs just has just ended up building a stack of proprietary software based around specific drives and hacked up existing applications, rather than attempting to modify any kind of modified firmware to just give what the laser sees, rather than a reinterpretation of it.  Maybe things have changed in the last 5 years, since I looked.

**Comment by 987123879113** [comment]
>As mentioned before, .cue is one of the worst 'non-standards' around, different programs ended up with different interpretations and layers of hacks, others followed suit. 

What is being discussed in this PR was all done used toc, not cue. I specifically went out of my way to avoid cues at all. This has nothing to do with cues.

I don't disagree with what you're saying about ripping but it's not something that I'm going to be able to solve and it feels like a dismissal of potentially real bugs. There is currently one real bug verified with real hardware (as fixed in the PR itself) and another that is being suggested based on real hardware testing and testing with other emulators that is potentially related. 

Any attempts to mention this issue on Discord so far has mostly been met with people saying "well did you try *this* solution to ripping the CDs?" which honestly isn't all that helpful and I don't want this PR to devolve into that too because I think there are real issues with the MAME code here if you don't try to hand wave it off as cue/ripping issues.

It could be that Duckstation has some kind of hack in place for this case. It could be that chdman is failing to parse tocs entirely correctly in this specific case (mixed mode CD, switching from data type to audio type which according to what I read in the standard *should* have a 150 gap). It could just be bad rips. Either way I think it deserves some investigation.

**Comment by DavidHaywood** [comment]
I agree, although TOC is just another representation of a high level, drive provided, interpretation of the information a specific drive provides, so is vulnerable to many of the same issues as cuesheets.

In the end it's something that needs dealing with, rather than blame shifting, so I guess basically what I'm saying is don't rule anything out, but be mindful of other cases too.

**Comment by cuavas** [comment]
Pinging @stenzek since comparisons with DuckStation are involved (feel free to ignore it though).

**Comment by stenzek** [comment]
> Unless you look at the duckstation code, I don't really know what it's doing. It might have its own hacks specifically to deal with these things.

I do actually have a hack in there for cues and a very specific case - audio track, multiple tracks in the same bin file, no index 0 specified => a two second pregap is automatically added between audio tracks. Got sick of people complaining that games were broken because of their old/bad dumps.

For CHDs, there's no such hack, it uses the metadata as-is.

**Comment by 987123879113** [comment]
I found some hacks for pgdatasize == 0 involving pregaps in cdrom.cpp. Removing those fixes all of the issues I encountered with the previously mentioned CHDs.

- The LBA returned by the CD tester program is 8775 which is the same a real PS1 and Duckstation
- The audio starts playing immediately instead of there being a pause before it plays for track 2 when playing in the PS1 BIOS CD player

Please check over the changes and tell me if this will actually break anything.

File: src/lib/util/cdrom.cpp [hunk 1]
```diff
@@ -455,7 +449,7 @@ chd_error read_partial_sector(cdrom_file *file, void *dest, uint32_t lbasector,
 	// if this is pregap info that isn't actually in the file, just return blank data
 	if (!phys)
 	{
-		if ((file->cdtoc.tracks[tracknum].pgdatasize == 0) && (lbasector < (file->cdtoc.tracks[tracknum].logframeofs + file->cdtoc.tracks[tracknum].pregap)))
+		if (file->cdtoc.tracks[tracknum].pgdatasize == 0 && lbasector < file->cdtoc.tracks[tracknum].logframeofs)
```
Comment (987123879113) [inline src/lib/util/cdrom.cpp:452]: The pregap is baked into logframeofs now so it shouldn't be added here. If it's added here then the audio will be silent during the beginning of the track. When the pregap section ends you can hear the sound suddenly cut in mid-track.

**Comment by cuavas** [comment]
Maybe hold off merging this until after freeze, just so it gets some shakedown time.  CDs are complicated things with lots of weirdness.

Philosophically, the pre-gap is part of the following track, not the precedeing track.  It’s just that the following track’s start point (i.e. the point where the TOC says the track starts) is at the end of the pre-gap.  You can see this if you look at the time display on a CD player for a conventionally mastered audio CD.  Skipping to a track jumps to the start point at the end of the pre-gap where the track elapsed time shows zero.  However, if you let the CD play from end to end, the track number rolls over as soon as it enters the pre-gap and it shows a negative elapsed time until it reaches the start point.

On a side note, there was a common “hidden track” technique that involved making the final track of the CD rather short, usually couple of seconds, with a song in the pre-gap.  This way, if you let the CD play normally, it would go straight from the last “normal” track to the hidden track and the negative elapsed time would count towards zero as the track played.  However, if you tried skipping to the hidden track, you’d jump straight to the silence following the pre-gap.  (You could skip to the hidden track, then seek backwards into its pre-gap to listen to it, too.)

To me, saying that you’re adding the pre-gap to the preceding track rings alarm bells – it disagrees with how things logically work from my experience mastering audio CDs.  However, reading the code, I think all this affects is the effective LBA-to-sector mapping, which is based on the TOC, so I think it’s OK.

**Comment by cuavas** [comment]
@rzero9 are you able to test this change on Japanese multimedia computers that use CD audio, and see what effects it has, positive or negative?

**Comment by DavidHaywood** [comment]
> On a side note, there was a common “hidden track” technique that involved making the final track of the CD rather short

The technique I'm more familiar with is putting a song before track 1 (rather than the final track) in the pregap there (similar to how CDI puts data there on some discs)

CD players would skip over the pregap, and you'd have to rewind to hear it. (not all CD players allowed it, some wouldn't let you 'rewind' into the pregap at all)

I own quite a few CDs like this, I seem to recall one has a pregap of about 20 minutes full of audio.

**Comment by 987123879113** [comment]
Just to be clear for anyone who wants to test out this PR: all of the changes in this PR are all gated behind a `pgdatasize == 0` check so testing CHDs where `pgdatasize` is 0 is required. Testing with non-0 `pgdatasize` CHDs is also welcome of course but testing with only non-0 `pgdatasize` CHDs could result in bugs going unnoticed.

I recommend uncommenting the long `printf("Track %02d is format ...` commands in both `cdrom_open` if you want to be sure.

**Comment by 987123879113** [comment]
I accidentally stumbled on a potentially similar issue while trying to debug why a game would work with a toc/bin compared to cue/bin. I think it's the same issue as what I fixed here, except it applies when pgdatasize is not 0.

I am using the same method to test with the PS1 BIOS CD player for audible testing, but the logs below are print statements directly from cdrom.cpp. This time I am loading the cue/bin and toc/bin directly from the file menu instead of converting it to CHD.

Working toc (presize/pgdatasize = 0):
```
Track 01 is format 1 subtype 2 datasize 2352 subsize 0 frames 8703 extraframes 0 pregap 0 pgmode 0 presize 0 postgap 0 logofs 0 physofs 0 chdofs 0
Track 02 is format 7 subtype 2 datasize 2352 subsize 0 frames 2683 extraframes 0 pregap 150 pgmode 0 presize 0 postgap 0 logofs 8853 physofs 8703 chdofs 0
Track 03 is format 7 subtype 2 datasize 2352 subsize 0 frames 1548 extraframes 0 pregap 17 pgmode 0 presize 0 postgap 0 logofs 11553 physofs 11386 chdofs 0
```

Broken cue (presize/pgdatasize != 0 so my PR changes don't apply here):
```
Track 01 is format 1 subtype 2 datasize 2352 subsize 0 frames 8853 extraframes 0 pregap 0 pgmode 0 presize 0 postgap 0 logofs 0 physofs 0 chdofs 0
Track 02 is format 7 subtype 2 datasize 2352 subsize 0 frames 2550 extraframes 0 pregap 0 pgmode 0 presize 0 postgap 0 logofs 8853 physofs 8853 chdofs 0
Track 03 is format 7 subtype 2 datasize 2352 subsize 0 frames 1541 extraframes 0 pregap 150 pgmode 7 presize 2352 postgap 0 logofs 11403 physofs 11403 chdofs 0
```

Track 2 plays correctly between both cue/bin and toc/bin. Track 3 plays correctly in the toc/bin set but is delayed in the cue/bin set.

Note that the logofs for track 1 and 2 are the same logofs for both sets, but the logofs of the broken cue version is 11403 instead of 11553 like in the toc version. 11403 + pregap (150) would put it at the same logofs as the toc version.

**Comment by rzero9** [comment]
> 
> 
> @rzero9 are you able to test this change on Japanese multimedia computers that use CD audio, and see what effects it has, positive or negative?

I did a quick test with Emit Vol. 1 and Vastness on the FM Towns, since they are extremely sensitive to changes in sector addressing (they play dialogue as small chunks of large CDDA tracks, starting at specific sectors, so if anything is wrong the characters will just talk gibberish). Everything seems fine so far, no audible differences at all.

**Comment by cuavas** [comment]
OK, now I’m confused.  The cue/bin and toc/bin are supposed to represent the same disc, right?  Parsing them correctly should produce functionally identical TOCs.  So then, why do they produce different CHDs?  There should be one correct CHD representation for a given CD, not different ones depending on the source format (unless one of the source formats is lossy or inadequate for representing the content of the disc).  I think that’s a major issue itself.

**Comment by 987123879113** [comment]
Sorry, I should've made it a little more clear. The results are were tested using this PR code. So the toc/bin, which has pgdatasize = 0, while the cue has pgdatasize != 0. My changes only apply when pgdatasize = 0 so the cue version doesn't get adjusted.

The images are ripped from the same disc. The cue version is ripped using CloneCD. The toc version is ripped using cdrdao. Both CHDs work in Duckstation as expected. The toc version works in MAME as expected. The cue version does not work as expected.

I will try later but I expect the toc version would have the same issue when not using the changes from this PR.

**Comment by cuavas** [comment]
So why is pgdatasize zero from one and not the other after being interpreted?  If they represent the same disk, surely it should end up with the same results?  What am I missing?

**Comment by rb6502** [comment]
As Haze alluded to, bin/cue has become more of a suggestion than a file format.  MAME follows the most common real-world behavior with it, but there can be bugs.  CDRDAO's bin/toc is much more rigidly defined.  I do agree that however this is resolved it should wait until after the release window.

**Comment by 987123879113** [comment]
>So why is pgdatasize zero from one and not the other after being interpreted? If they represent the same disk, surely it should end up with the same results? What am I missing?

When talking strictly about loading from cue and toc:

The cue parser will set pgdatasize when `INDEX 01` is encountered: https://github.com/mamedev/mame/blob/master/src/lib/util/chdcd.cpp#L981-L1032

The toc parser will never set pgdatasize to anything.


When loading from CHD:
`pgdatasize` will be set when `cdrom_convert_type_string_to_pregap_info` is called, which will only occur when the pregap value is greater than 0 and if `pgtype` starts with "V": https://github.com/mamedev/mame/blob/master/src/lib/util/cdrom.cpp#L1225-L1228

`pgtype` will start with "V" in a CHD only when `pgdatasize` was greater than 0 when creating the CHD. So I don't think there will ever be CHDs made from tocs that will have `pgdatasize` greater than 0 if I'm understanding the code correctly, since only cue will set `pgdatasize` to non-0.
https://github.com/mamedev/mame/blob/master/src/lib/util/cdrom.cpp#L1316-L1324


>I do agree that however this is resolved it should wait until after the release window.
I mentioned it to cuavas elsewhere but I am also in agreement that this PR should wait for more discussion and testing before being merged.

**Comment by 987123879113** [comment]
@rzero9 Would it be possible for you to try this specific patch with those same games? I made some more invasive changes but I wasn't able to get PC-98 stuff working so I couldn't test the games you mentioned. All of the previously broken games I personally wanted to test (System 573 and Firebeat games) are fixed with this patch but it has a higher chance to break other things so I haven't pushed the commit directly to this PR yet.

https://github.com/987123879113/mame/commit/e4b8fe1366a543660639df564e0c7d769d4940b5.patch

**Comment by rzero9** [comment]
Unfortunately it does seem to cause issues this time. I tested Emit Vol. 1 on FM Towns and Alone in the Dark 2 on PC-98, and it's playing the tracks a couple of seconds ahead of where it should be. Testing anything on PC-98 is hard, though, as there seems to have been a regression at some point and writing to floppy disks (which is needed for a lot of CD games) doesn't work very well.

You can see for yourself how the audio desyncs in Emit here: https://youtu.be/8D2DyS8RzKY

Keep in mind the games I tested are CHDs converted from Redump cuesheets, which always have "real" pregaps instead of generated ones (INDEX 00 command instead of PREGAP). The rest of the entries in both softlists come from random/unknown sources, so I don't trust them too much to be correct.

**Comment by 987123879113** [comment]
@rzero9 Thanks for testing, I was finally able to get Emit Vol 1 working. I was trying to boot the FM Towns data with PC-98. :innocent: 
The CHD checksum and everything matches what's in the database as well so it should be a good dump for testing. And I also get the exact same issue shown in your video now.

I looked at the FM Towns code and I think I see why this problem is occurring. In the [fmtowns: cdrom toc fixes (nw)](https://github.com/mamedev/mame/commit/81f44e1e968b559330a0c9cefebcef0868e1e9f4) commit from a few years ago, they accounted for the pregap in the FM Towns CD code already: `addr += cdrom_get_toc(m_cdrom->get_cdrom_file())->tracks[track].pregap;`.

https://github.com/mamedev/mame/blob/master/src/mame/drivers/fmtowns.cpp#L1840

Removing that live from fmtowns.cpp fixes the audio again, since pregap is accounted for in cdrom.cpp now in my patch.

**Comment by 987123879113** [comment]
I pushed the new code direcetly into the PR with some additional fixes and renamed the PR to more accurately reflect the changes.

- FM Towns used `cdrom_get_toc` to adjust for the pregap in its own CD-ROM code, so I am searching for other similar instances throughout MAME
- The PSX CD player tries to play the full length of the track specified by frames, which is calculated by `outinfo.track[trknum+1].idx0offs - outinfo.track[trknum].idx0offs` and will also include pregap. This results in silence being played for a few seconds when the actual audio is finished playing but it hasn't realized it finished yet. I tried to adjust for that. Will require some more testing.
- Mattel Hyperscan (spg290_cdservo.cpp) does a bunch of stuff with the generated toc when loading CDs returned by `cdrom_get_toc` but I couldn't find any Hyperscan dumps that actually have more than one track or anything relating to pregap, so Hyperscan should work the same as before.

Some other targets I want to test still: gdrom.cpp, pce_cd.cpp, megacdcd.cpp, and try to debug what's going on with Alone in the Dark 2 for PC-98 if I can get it working.

Please let me know if anything I haven't mentioned is broken by these changes and I'll take a look at it.

**Pull request merged**